src/VecSim/algorithms/svs/svs.h

```diff
@@ -0,0 +1,454 @@
+                                 : params_.epsilon;
+
+        const auto range_search_boundaries = radius * (1.0 + epsilon);
+        bool keep_searching = true;
+        while (keep_searching && svs_it.size() > 0) {
+            for (auto &neighbor : svs_it) {
+                const auto dist = toVecSimDistance(neighbor.distance());
+                if (dist <= radius) {
+                    rep->results.emplace_back(neighbor.id(), dist);
+                } else if (dist > range_search_boundaries) {
+                    keep_searching = false;
+                    break;
+                }
+            }
+            if (keep_searching) {
+                svs_it.next(cancel);
+                if (cancel()) {
+                    rep->code = VecSim_QueryReply_TimedOut;
+                    return rep;
+                }
+            }
+        }
+#else
+        // strict range search assuming that batch iterator results are not sorted in 100%
+        int batch_times = 3;
+        bool done = false;
+        while (svs_it.size() > 0 && batch_times > 0) {
+            for (auto &neighbor : svs_it) {
+                if (toVecSimDistance(neighbor.distance()) <= radius) {
+                    rep->results.emplace_back(neighbor.id(), toVecSimDistance(neighbor.distance()));
+                    done = false;
+                } else {
+                    done = true;
+                }
+            }
+            if (done)
+                if (--batch_times == 0)
+                    break;
+            svs_it.next(cancel);
+            if (cancel()) {
+                rep->code = VecSim_QueryReply_TimedOut;
+                return rep;
+            }
+        }
+#endif
+        return rep;
+    }
+
+    VecSimBatchIterator *newBatchIterator(const void *queryBlob,
+                                          VecSimQueryParams *queryParams) const override {
+        auto *queryBlobCopy =
+            this->allocator->allocate_aligned(this->dataSize, this->preprocessors->getAlignment());
+        memcpy(queryBlobCopy, queryBlob, this->dim * sizeof(DataType));
+        this->preprocessQueryInPlace(queryBlobCopy);
+        // Ownership of queryBlobCopy moves to VecSimBatchIterator that will free it at the end.
+        if (indexSize() == 0) {
+            return new (this->getAllocator())
+                NullSVS_BatchIterator(queryBlobCopy, queryParams, this->getAllocator());
+        } else {
+            // TODO control impl_ lifetime (shared_ptr?)
+            return new (this->getAllocator()) SVS_BatchIterator<impl_type, data_type>(
+                queryBlobCopy, impl_, queryParams, this->getAllocator());
+        }
+    }
+
+    bool preferAdHocSearch(size_t subsetSize, size_t k, bool initial_check) const override {
+        bool res = true;
+        this->lastMode =
+            res ? (initial_check ? HYBRID_ADHOC_BF : HYBRID_BATCHES_TO_ADHOC_BF) : HYBRID_BATCHES;
+        return res;
+    }
```

Comment:
This is not implemented... We should come up with some basic heuristics that tell us what is the preferable approach to take in case of a filtered query based on the number of elements that are assumed to pass the filter (`subsetSize`), `k`, and index properties. This will have a significant performance impact...

Reply:
Basic heuristic has been implemented. Please review.

Reply:
Looks good